File: sdks/ruby/lib/sequins_otel.rb
# frozen_string_literal: true

require "opentelemetry-sdk"
require "opentelemetry-logs-sdk"
require "opentelemetry-exporter-otlp"
require "opentelemetry-exporter-otlp-logs"
require_relative "sequins_otel/version"

# SequinsOtel — OpenTelemetry distro for Sequins.
#
# Zero-config observability for local development. Configures traces, metrics, and logs
# to export to a local Sequins instance via OTLP/HTTP.
#
# @example Basic usage
#   require "sequins_otel"
#
#   sequins = SequinsOtel.init(service_name: "my-app")
#   tracer = sequins.tracer_provider.tracer("my-module")
#
#   # On exit:
#   sequins.shutdown
#
module SequinsOtel
  DEFAULT_ENDPOINT = "http://localhost:4318"

  # Returned by {SequinsOtel.init}. Provides access to all configured providers.
  class Handle
    attr_reader :tracer_provider, :meter_provider, :logger_provider

    def initialize(tracer_provider:, meter_provider:, logger_provider:)
      @tracer_provider = tracer_provider
      @meter_provider = meter_provider
      @logger_provider = logger_provider
    end

    # Flushes all pending telemetry and shuts down all providers.
    # Always call this before your process exits.
    def shutdown
      @tracer_provider&.shutdown
      @meter_provider&.shutdown
      @logger_provider&.shutdown
    end
  end

  # Initialize OpenTelemetry with all signals pre-configured for Sequins.
  #
  # @param service_name [String] the name of your service
  # @param endpoint [String] the OTLP HTTP endpoint (default: http://localhost:4318)
  # @param metric_export_interval [Integer] metric export interval in milliseconds (default: 10000)
  # @param use_all [Boolean] when true, installs all available OTel instrumentations
  #   (requires the +opentelemetry-instrumentation-all+ gem). Install it and specific
  #   framework gems you want, e.g. +gem "opentelemetry-instrumentation-rack"+.
  # @return [Handle] a handle providing access to all configured providers
  def self.init(
    service_name: nil,
    endpoint: nil,
    metric_export_interval: 10_000,
    use_all: false
  )
    resolved_service_name = service_name ||
      ENV["OTEL_SERVICE_NAME"] ||
      "unknown_service"

    resolved_endpoint = endpoint ||
      ENV["OTEL_EXPORTER_OTLP_ENDPOINT"] ||
      DEFAULT_ENDPOINT

    resource = OpenTelemetry::SDK::Resources::Resource.create(
      "service.name" => resolved_service_name
    )

    exporter = OpenTelemetry::Exporter::OTLP::Exporter.new(
      endpoint: "#{resolved_endpoint}/v1/traces",
      compression: "gzip"
    )

    # Configure the global SDK
    OpenTelemetry::SDK.configure do |c|
      c.resource = resource
      c.add_span_processor(
        OpenTelemetry::SDK::Trace::Export::BatchSpanProcessor.new(exporter)
      )
      if use_all
        begin
          require "opentelemetry-instrumentation-all"
          c.use_all
        rescue LoadError
          # opentelemetry-instrumentation-all not installed — skipping auto-instrumentation.
          # Add it to your Gemfile: gem "opentelemetry-instrumentation-all"
        end
      end
    end

    tracer_provider = OpenTelemetry.tracer_provider
    meter_provider = OpenTelemetry.meter_provider rescue nil

    # --- Logs ---
    log_exporter = OpenTelemetry::Exporter::OTLP::Logs::OtlpHttpLogRecordExporter.new(
      endpoint: "#{resolved_endpoint}/v1/logs",
      compression: "gzip"
    )
    logger_provider = OpenTelemetry::SDK::Logs::LoggerProvider.new(resource: resource)
    logger_provider.add_log_record_processor(
      OpenTelemetry::SDK::Logs::Export::BatchLogRecordProcessor.new(log_exporter)
    )
    OpenTelemetry::Logs.logger_provider = logger_provider

    Handle.new(
      tracer_provider: tracer_provider,
      meter_provider: meter_provider,
      logger_provider: logger_provider
    )
  end
end

File: sdks/elixir/lib/sequins_otel.ex
defmodule SequinsOtel do
  @moduledoc """
  OpenTelemetry distro for Sequins — zero-config observability for local development.

  Configures OpenTelemetry traces, metrics, and logs to export to a local Sequins
  instance via OTLP/gRPC.

  ## Usage

      # In your application.ex:
      def start(_type, _args) do
        SequinsOtel.configure(service_name: "my-app")
        # ... start your supervision tree
      end

  Or use the `init/1` function which also starts the OTel application:

      {:ok, _} = SequinsOtel.init(service_name: "my-app")

  ## Environment variables

  - `OTEL_SERVICE_NAME` — overrides `service_name`
  - `OTEL_EXPORTER_OTLP_ENDPOINT` — overrides `endpoint`
  """

  @default_endpoint "http://localhost:4317"

  @doc """
  Configures OpenTelemetry application environment for Sequins.

  This is the preferred approach — call it before starting the OTel application,
  typically in your `Application.start/2` callback.

  ## Options

    * `:service_name` — service name (default: `OTEL_SERVICE_NAME` env var or `"unknown_service"`)
    * `:endpoint` — OTLP gRPC endpoint (default: `OTEL_EXPORTER_OTLP_ENDPOINT` env var or `"http://localhost:4317"`)
    * `:traces_enabled` — enable trace export (default: `true`)
    * `:metrics_enabled` — enable metrics export (default: `true`)
    * `:logs_enabled` — enable log export (default: `true`)
    * `:auto_instrument` — when `true`, calls `.setup()` on available instrumentation
      packages (`opentelemetry_phoenix`, `opentelemetry_ecto`, `opentelemetry_oban`).
      Each must be added separately to your `mix.exs`. (default: `false`)
  """
  @spec configure(keyword()) :: :ok
  def configure(opts \\ []) do
    service_name =
      Keyword.get(opts, :service_name) ||
        System.get_env("OTEL_SERVICE_NAME") ||
        "unknown_service"

    endpoint =
      Keyword.get(opts, :endpoint) ||
        System.get_env("OTEL_EXPORTER_OTLP_ENDPOINT") ||
        @default_endpoint

    # Configure resource
    Application.put_env(:opentelemetry, :resource, [
      service: [name: service_name]
    ])

    # Configure OTLP exporter
    Application.put_env(:opentelemetry_exporter, :otlp_protocol, :grpc)
    Application.put_env(:opentelemetry_exporter, :otlp_endpoint, endpoint)

    # Configure batch span processor
    Application.put_env(:opentelemetry, :processors, [
      otel_batch_processor: %{
        exporter: {
          :opentelemetry_exporter,
          %{endpoints: [endpoint]}
        }
      }
    ])

    if Keyword.get(opts, :auto_instrument, false) do
      _try_setup(OpentelemetryPhoenix)
      _try_setup(OpentelemetryEcto)
      _try_setup(OpentelemetryOban)
    end

    :ok
  end

  defp _try_setup(module) do
    if Code.ensure_loaded?(module), do: module.setup()
  end

  @doc """
  Initializes OpenTelemetry for Sequins and starts the OTel application.

  Returns `{:ok, sequins}` where `sequins` is a map with the configured settings.
  Call `SequinsOtel.shutdown/1` when your application exits.

  Accepts all options from `configure/1`, including `:auto_instrument`.

  ## Example

      {:ok, sequins} = SequinsOtel.init(service_name: "my-app")
      # Use :opentelemetry_api to create spans:
      # require OpenTelemetry.Tracer
      # OpenTelemetry.Tracer.with_span "my-operation" do ... end

      # On exit:
      SequinsOtel.shutdown(sequins)
  """
  @spec init(keyword()) :: {:ok, map()} | {:error, term()}
  def init(opts \\ []) do
    :ok = configure(opts)

    case Application.ensure_all_started(:opentelemetry) do
      {:ok, _apps} ->
        service_name =
          Keyword.get(opts, :service_name) ||
            System.get_env("OTEL_SERVICE_NAME") ||
            "unknown_service"

        endpoint =
          Keyword.get(opts, :endpoint) ||
            System.get_env("OTEL_EXPORTER_OTLP_ENDPOINT") ||
            @default_endpoint

        {:ok, %{service_name: service_name, endpoint: endpoint}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Shuts down OpenTelemetry, flushing any pending spans.
  """
  @spec shutdown(map()) :: :ok
  def shutdown(_sequins) do
    :opentelemetry.shutdown_tracer_provider(:global)
    :ok
  end
end
